src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs

```diff
@@ -189,32 +280,61 @@ public void Dispose()
 
         internal AbstractProject GetProject(ProjectId id)
         {
-            AbstractProject project;
-            _projectMap.TryGetValue(id, out project);
-            return project;
+            lock (_gate)
+            {
+                AbstractProject project;
+                _projectMap.TryGetValue(id, out project);
+                return project;
+            }
         }
 
         /// <summary>
         /// Add a project to the workspace.
+        /// If invoked on the foreground thread, the add is executed right away.
+        /// Otherwise, the add is scheduled on foreground task scheduler.
```

Comment:
I guess my question is more "why can't I just add projects on a background thread? why are we moving to the foreground thread to notify workspace hosts?"